src/Raven.Server/Documents/Handlers/DocumentReplicationHandler.cs

```diff
@@ -50,110 +47,113 @@ private long GetLastEtag(Guid srcDbId, DocumentsOperationContext context)
+                                        messageTypeAsString,
+                                        message,
+                                        webSocketStream,
+                                        context,
+                                        srcDbId,
+                                        incomingReplication);
                                 }
-                                executer.ReceiveReplicatedDocuments(context, docs);
-
-                                //precaution
-                                if (tx == null)
-                                    throw new InvalidOperationException(@"
-                                        Transaction is not initialized while receiving replicated documents; this
-                                         is something that is not supposed to happen and is likely a bug.");
-
-                                tx.Commit();
                             }
                             catch (Exception e)
                             {
                                 throw new InvalidOperationException($@"Failed to receive replication document batch. (Origin -> Database Id = {srcDbId}, Database Name = {srcDbName}, Origin URL = {srcUrl})", e);
                             }
-                            finally
-                            {
-                                tx?.Dispose();
-                            }
-
-                            if (hasOtherSideClosedConnection)
-                                break;
                         }
                     }
-
-                    //if execution path gets here, it means the node was disconnected
-                    Database.DocumentReplicationLoader.HandleConnectionDisconnection(executer);
                 }
             }
         }
 
-        private void WriteLastEtagResponse(WebSocket webSocket, DocumentsOperationContext context, Guid srcDbId)
+        private void HandleMessage(string messageTypeAsString, 
+            BlittableJsonReaderObject message, 
+            WebsocketStream webSocketStream, 
+            DocumentsOperationContext context, 
+            Guid srcDbId, 
+            IncomingDocumentReplication incomingReplication)
         {
-            using (var websocketStream = new WebsocketStream(webSocket, Database.DatabaseShutdown))
-            using (var responseWriter = new BlittableJsonTextWriter(context, websocketStream))
+            switch (messageTypeAsString)
+            {
+                case Constants.Replication.MessageTypes.GetLastEtag:
+                    using (context.OpenReadTransaction())
+                        WriteLastEtagResponse(webSocketStream, context, srcDbId);
+                    break;
+                case Constants.Replication.MessageTypes.ReplicationBatch:
+                    BlittableJsonReaderArray replicatedDocs;
+                    if (!message.TryGet(Constants.Replication.PropertyNames.ReplicationBatch, out replicatedDocs))
```

Comment:
Love the errors, but where are they handled? They need to be sent to the other side and logged.